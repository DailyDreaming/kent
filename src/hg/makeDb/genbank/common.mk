#Optimized for performance or debugging
#OPT= -ggdb
#OPT= -ggdb -O3
OPT= -ggdb -O

KENT = ${GBROOT}/../../..

# FIXME: for now, need to link statically on RH7 or a warning is written
# to stdout on RH9, which breaks the program trying to read the output. 
# Gag me...
ifneq ($(wildcard ${GBROOT}/extern/lib/libmysqlclient.a),)
MYSQLLIBS=${GBROOT}/extern/lib/libmysqlclient.a
STATIC = -static
endif

ifeq (${MYSQLLIBS},)
$(error must set MYSQLLIBS env var)
endif

INCL = -I${GBROOT}/src/inc -I${KENT}/inc -I${KENT}/hg/inc
CFLAGS = ${OPT} ${STATIC} -DJK_WARN -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_GNU_SOURCE -Wall -Werror ${INCL}

BINDIR = ${GBROOT}/bin
BINARCH = ${BINDIR}/${MACHTYPE}
LIBDIR = ${GBROOT}/lib
LIBARCH = ${LIBDIR}/${MACHTYPE}

LIBGENBANK = $(LIBARCH)/libgenbank.a

MYLIBDIR = ${KENT}/lib/$(MACHTYPE)
LIBS = $(LIBGENBANK) $(MYLIBDIR)/jkhgap.a $(MYLIBDIR)/jkweb.a ${MYSQLLIBS} -lm

TESTBIN = ${GBROOT}/tests/bin
TESTBINARCH = ${TESTBIN}/$(MACHTYPE)

MKDIR = mkdir -p

%.o: %.c
	${CC} ${CFLAGS} -c -o $@ $<

$(BINARCH)/%: ${O} makefile ${LIBGENBANK}
	@${MKDIR} -p ${BINARCH}
	gcc ${CFLAGS} -o $@ $O $(LIBS)

${BINDIR}/%: %
	@${MKDIR} -p ${BINDIR}
	cp -f $< $@
	chmod a-w,a+rx $@
