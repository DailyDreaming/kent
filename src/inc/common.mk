CC=gcc
COPT=-O 
CFLAGS=
HG_DEFS=-D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_GNU_SOURCE -DMACHTYPE_${MACHTYPE}
HG_WARN=-Wformat -Wimplicit -Wuninitialized -Wreturn-type
HG_INC=-I../inc -I../../inc -I../../../inc -I../../../../inc
MACHTYPE=osx
# add the follow to makefiles to enable stronger warning checks
# HG_WARN = ${HG_WARN_ERR}
HG_WARN_ERR = -DJK_WARN -Wall -Werror

CGI_BIN=/usr/local/apache/cgi-bin
MKDIR=mkdir -p

.c.o:
	${CC} ${COPT} ${CFLAGS} ${HG_DEFS} ${HG_WARN} ${HG_INC} ${XINC} -c $*.c


