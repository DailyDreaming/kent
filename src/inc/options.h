/* Stuff to process options out of command line. */

#ifndef OPTIONS_H

#ifndef HASH_H
#include "hash.h"
#endif

char *optionVal(char *name, char *defaultVal);
/* Return named option if in options hash, otherwise default. */

int optionInt(char *name, int defaultVal);
/* Return integer value of named option, or default value
 * if not set. */

float optionFloat(char *name, float defaultVal);
/* Return float value of named option or default. */

boolean optionExists(char *name);
/* Return TRUE if option has been set. */

void optionHash(int *pArgc, char *argv[]);
/* Read options in command line into options hash.   
 * Options come in three forms:
 *      -option         words starting with dash
 *      option=val      words with = in the middle
 *      -option=val     combining the two.
 * The resulting hash will be keyed by the option name with the val
 * string for value.  For '-option' types the value is 'on'. */

void optionHashSome(int *pArgc, char *argv[], boolean justFirst);
/* Set up option hash from command line, optionally only adding
 * up to first non-optional word. */

struct hash *optionParseIntoHash(int *pArgc, char *argv[], boolean justFirst);
/* Read options in argc/argv into a hash of your own choosing. */

#endif /* OPTIONS_H */

