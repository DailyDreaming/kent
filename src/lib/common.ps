%% Common postScript header.

/boxHere {
% Draw a box at current position.
% Usage width height boxAt
dup 0 exch   %stack width height 0 height
rlineto      %stack width height
exch dup 0   %stack height width width 0
rlineto      %stack height width
exch         %stack width height
0 exch       %stack width 0 height
neg          %stack width 0 -height
rlineto      %stack width
neg 0        %stack -width 0
rlineto
} def

/boxAt {
% Draw a box at a position.  Usage:
% width height x y boxAt
moveto
boxHere
} def

/fillBox {
% Draw filled box at a position. Usage:
% width height x y fillBox
newpath
boxAt
fill
} def

/strokeBox {
% Draw stroked box at a position. Usage:
% width height x y strokeBox
newpath
boxAt
stroke
} def

/smallFont {
/Times-Roman findfont
10 scalefont
setfont
} def

smallFont
