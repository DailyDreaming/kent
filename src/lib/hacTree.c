/* hacTree - Hierarchical Agglomerative Clustering a list of inputs into a binary tree */

#include "common.h"
#include "hacTree.h"

static struct hacTree *leafNodesFromItems(const struct slList *itemList, int itemCount,
					  struct lm *localMem)
/* Allocate & initialize leaf nodes that contain only items. */
{
struct hacTree *leafNodes = lmAlloc(localMem, itemCount * sizeof(struct hacTree));
int i = 0;
const struct slList *item = itemList;
while (item != NULL && i < itemCount)
    {
    // needMem zeroes the memory, so initialize only non-NULL stuff.
    struct hacTree *node = &(leafNodes[i]);
    if (i < itemCount-1)
	node->next = &(leafNodes[i+1]);
    node->itemOrCluster = (struct slList *)item;
    i++;
    item = item->next;
    }
return leafNodes;
}

struct sortWrapper
/* We need to compare nodes' itemOrClusters using cmpF and extraData;
 * qsort's comparison function doesn't have a way to pass in extraData,
 * so we need to point to it from each qsort element. */
{
    struct hacTree *node;  // contains itemOrCluster to be compared
    hacCmpFunction *cmpF;  // user-provided itemOrCluster comparison function
    void *extraData;       // user-provided aux data for cmpF
};

static int sortWrapCmp(const void *v1, const void *v2)
/* Unpack sortWrappers and run cmpF on nodes' itemOrClusters with extraData. */
{
const struct sortWrapper *w1 = v1, *w2 = v2;
return w1->cmpF(w1->node->itemOrCluster, w2->node->itemOrCluster, w1->extraData);
}

static struct sortWrapper *makeSortedWraps(struct hacTree *leafNodes, int itemCount,
					   struct lm *localMem, hacCmpFunction cmpF,
					   void *extraData)
/* Use cmpF and extraData to sort wrapped leaves so that identical leaves will be adjacent. */
{
struct sortWrapper *leafWraps = lmAlloc(localMem, itemCount * sizeof(struct sortWrapper));
int i;
for (i=0;  i < itemCount;  i++)
    {
    leafWraps[i].node = &(leafNodes[i]);
    leafWraps[i].cmpF = cmpF;
    leafWraps[i].extraData = extraData;
    }
qsort(leafWraps, itemCount, sizeof(struct sortWrapper), sortWrapCmp);
return leafWraps;
}

INLINE void initNode(struct hacTree *node, const struct hacTree *left, const struct hacTree *right,
		     hacDistanceFunction *distF, hacMergeFunction *mergeF, void *extraData)
/* Initialize node to have left and right as its children.  Leave parent pointers
 * alone -- they would be unstable during tree construction. */
{
node->left = (struct hacTree *)left;
node->right = (struct hacTree *)right;
node->childDistance = distF(left->itemOrCluster, right->itemOrCluster, extraData);
node->itemOrCluster = mergeF(left->itemOrCluster, right->itemOrCluster, extraData);
}

INLINE struct hacTree preClusterNodes(const struct sortWrapper *leafWraps, int i, int runLength,
				      hacDistanceFunction *distF, hacMergeFunction *mergeF,
				      void *extraData, struct lm *localMem)
/* Caller has allocated a node, and this returns what to store there:
 * a recursively constructed cluster of nodes extracted from wrapped
 * leafNodes (leafWraps) starting at i, for runLength items. */
{
struct hacTree ret;
if (runLength > 2)
    {
    struct hacTree *newClusters = lmAlloc(localMem, 2 * sizeof(struct hacTree));
    int halfLength = runLength/2;
    newClusters[0] = preClusterNodes(leafWraps, i, halfLength,
				     distF, mergeF, extraData, localMem);
    newClusters[1] = preClusterNodes(leafWraps, i+halfLength, runLength-halfLength,
				     distF, mergeF, extraData, localMem);
    initNode(&ret, &(newClusters[0]), &(newClusters[1]), distF, mergeF, extraData);
    }
else if (runLength == 2)
    {
    initNode(&ret, leafWraps[i].node, leafWraps[i+1].node, distF, mergeF, extraData);
    }
else
    ret = *(leafWraps[i].node);
return ret;
}

static struct hacTree *sortAndPreCluster(struct hacTree *leafNodes, int *retItemCount,
					 struct lm *localMem, hacDistanceFunction *distF,
					 hacMergeFunction *mergeF, hacCmpFunction *cmpF,
					 void *extraData)
/* Use cmpF and extraData to sort wrapped leaf nodes so that identical leaves will be adjacent,
 * then replace leaves with clusters of identical leaves where possible.  Place new
 * (hopefully smaller) item count in retItemCount. */
{
int itemCount = *retItemCount;
struct sortWrapper *leafWraps = makeSortedWraps(leafNodes, itemCount, localMem, cmpF, extraData);
struct hacTree *newLeaves = lmAlloc(localMem, itemCount * sizeof(struct hacTree));
int i=0, newI=0;
while (i < itemCount)
    {
    int nextRunStart;
    for (nextRunStart = i+1;  nextRunStart < itemCount; nextRunStart++)
	if (distF(leafWraps[i].node->itemOrCluster, leafWraps[nextRunStart].node->itemOrCluster,
		  extraData) != 0)
	    break;
    int runLength = nextRunStart - i;
    newLeaves[newI] = preClusterNodes(leafWraps, i, runLength, distF, mergeF, extraData, localMem);
    i = nextRunStart;
    newI++;
    }
*retItemCount = newI;
return newLeaves;
}

static struct hacTree *pairUpItems(const struct slList *itemList, int itemCount,
				   int *retPairCount, struct lm *localMem,
				   hacDistanceFunction *distF, hacMergeFunction *mergeF,
				   hacCmpFunction *cmpF, void *extraData)
/* Allocate & initialize leaf nodes and all possible pairings of leaf nodes
 * which will be our seed clusters.  If cmpF is given, pre-sort the leaf nodes
 * and pre-cluster identical leaves before generating seed clusters. */
{
struct hacTree *leafNodes = leafNodesFromItems(itemList, itemCount, localMem);
if (cmpF != NULL)
    leafNodes = sortAndPreCluster(leafNodes, &itemCount, localMem,
				  distF, mergeF, cmpF, extraData);
int pairCount = itemCount * (itemCount-1) / 2;
struct hacTree *pairHeap = lmAlloc(localMem, pairCount * sizeof(struct hacTree));
int i, j, pairIx;
for (i=0, pairIx=0;  i < itemCount-1;  i++)
    for (j=i+1;  j < itemCount;  j++, pairIx++)
	initNode(&(pairHeap[pairIx]), &(leafNodes[i]), &(leafNodes[j]), distF, mergeF, extraData);
*retPairCount = pairCount;
return pairHeap;
}

struct hacTree *hacTreeFromItems(const struct slList *itemList, struct lm *localMem,
				 hacDistanceFunction *distF, hacMergeFunction *mergeF,
				 hacCmpFunction *cmpF, void *extraData)
/* Using distF, mergeF, optionally cmpF and binary tree operations,
 * perform a hierarchical agglomerative (bottom-up) clustering of
 * items.  To free the resulting tree, lmCleanup(&localMem). */
{
if (itemList == NULL)
    return NULL;
struct hacTree *root = NULL;
int itemCount = slCount(itemList);
int pairCount = 0;
struct hacTree *leafPairs = pairUpItems(itemList, itemCount, &pairCount, localMem,
					distF, mergeF, cmpF, extraData);
struct hacTree *heapHead = leafPairs;
int heapLength = pairCount;
while (heapLength > 0)
    {
    // Scan heap for node with lowest childDistance; swap that node w/head
    int i;
    int bestIx = 0;
    double minScore = heapHead[0].childDistance;
    for (i=1;  i < heapLength;  i++)
	if (heapHead[i].childDistance < minScore)
	    {
	    minScore = heapHead[i].childDistance;
	    bestIx = i;
	    }
    if (bestIx != 0)
	swapBytes((char *)&(heapHead[0]), (char *)&(heapHead[bestIx]), sizeof(struct hacTree));
    // Pop the best (lowest-distance) node from heapHead, make it root (for now).
    root = heapHead;
    heapHead = &(heapHead[1]);
    heapLength--;
    // Where root->left is found in the heap, replace it with root.
    // Where root->right is found, drop that node so it doesn't become
    // a duplicate of the replacement cases.
    for (i=0;  i < heapLength;  i++)
	{
	struct hacTree *node = &(heapHead[i]);
	if (node->left == root->left)
	    initNode(node, root, node->right, distF, mergeF, extraData);
	else if (node->right == root->left)
	    initNode(node, root, node->left, distF, mergeF, extraData);
	else if (node->left == root->right || node->right == root->right)
	    {
	    if (i < heapLength-1)
		memcpy(node, &(heapHead[i+1]), (heapLength-i-1)*sizeof(struct hacTree));
	    heapLength--;
	    i--;
	    }
	}
    // root now has a stable address, unlike nodes still in the heap, so set parents here:
    if (root->left != NULL)
	root->left->parent = root;
    if (root->right != NULL)
	root->right->parent = root;
    }
// This shouldn't be necessary as long as initNode leaves parent pointers alone,
// but just in case that changes:
root->parent = NULL;
return root;
}
