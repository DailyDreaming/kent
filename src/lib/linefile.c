/*****************************************************************************
 * Copyright (C) 2000 Jim Kent.  This source code may be freely used         *
 * for personal, academic, and non-profit purposes.  Commercial use          *
 * permitted only by explicit agreement with Jim Kent (jim_kent@pacbell.net) *
 *****************************************************************************/
/* lineFile - stuff to rapidly read text files and parse them into
 * lines. */

#include "common.h"
#include <fcntl.h>
#include "linefile.h"

struct lineFile *lineFileAttatch(char *fileName, bool zTerm, int fd)
/* Wrap a line file around an open'd file. */
{
struct lineFile *lf;
AllocVar(lf);
lf->fileName = cloneString(fileName);
lf->fd = fd;
lf->bufSize = 256*1024;
lf->zTerm = zTerm;
lf->buf = needMem(lf->bufSize+1);
return lf;
}

void lineFileExpandBuf(struct lineFile *lf, int newSize)
/* Expand line file buffer. */
{
assert(newSize > lf->bufSize);
lf->buf = needMoreMem(lf->buf, lf->bytesInBuf, newSize);
lf->bufSize = newSize;
}


struct lineFile *lineFileStdin(bool zTerm)
/* Wrap a line file around stdin. */
{
return lineFileAttatch("stdin", zTerm, fileno(stdin));
}

struct lineFile *lineFileMayOpen(char *fileName, bool zTerm)
/* Try and open up a lineFile. */
{
int fd;

if (sameString(fileName, "stdin"))
    return lineFileStdin(zTerm);
fd = open(fileName, O_RDONLY);
if (fd == -1)
    return NULL;
return lineFileAttatch(fileName, zTerm, fd);
}

struct lineFile *lineFileOpen(char *fileName, bool zTerm)
/* Open up a lineFile or die trying. */
{
struct lineFile *lf = lineFileMayOpen(fileName, zTerm);
if (lf == NULL)
    errAbort("Couldn't open %s", fileName);
return lf;
}

void lineFileReuse(struct lineFile *lf)
/* Reuse current line. */
{
lf->reuse = TRUE;
}

int lineFileLongNetRead(int fd, char *buf, int size)
/* Keep reading until either get no new characters or
 * have read size */
{
int oneSize, totalRead = 0;

while (size > 0)
    {
    oneSize = read(fd, buf, size);
    if (oneSize <= 0)
        break;
    totalRead += oneSize;
    buf += oneSize;
    size -= oneSize;
    }
return totalRead;
}

boolean lineFileNext(struct lineFile *lf, char **retStart, int *retSize)
/* Fetch next line from file. */
{
char *buf = lf->buf;
int bytesInBuf = lf->bytesInBuf;
int endIx;
boolean gotLf = FALSE;
int newStart;

if (lf->reuse)
    {
    lf->reuse = FALSE;
    if (retSize != NULL)
	*retSize = lf->lineEnd - lf->lineStart;
    *retStart = buf + lf->lineStart;
    return TRUE;
    }
/* Find next end of line in buffer. */
for (endIx = lf->lineEnd; endIx < bytesInBuf; ++endIx)
    {
    if (buf[endIx] == '\n')
	{
	gotLf = TRUE;
	endIx += 1;
	break;
	}
    }

/* If not in buffer read in a new buffer's worth. */
while (!gotLf)
    {
    int oldEnd = lf->lineEnd;
    int sizeLeft = bytesInBuf - oldEnd;
    int bufSize = lf->bufSize;
    int readSize = bufSize - sizeLeft;

    if (oldEnd > 0 && sizeLeft > 0)
	memmove(buf, buf+oldEnd, sizeLeft);
    lf->bufOffsetInFile += oldEnd;
    readSize = lineFileLongNetRead(lf->fd, buf+sizeLeft, readSize);
    if (readSize + sizeLeft <= 0)
	{
	lf->bytesInBuf = lf->lineStart = lf->lineEnd = 0;
	return FALSE;
	}
    bytesInBuf = lf->bytesInBuf = readSize + sizeLeft;
    lf->lineEnd = 0;

    /* Look for next end of line.  */
    for (endIx = sizeLeft; endIx <bytesInBuf; ++endIx)
	{
	if (buf[endIx] == '\n')
	    {
	    endIx += 1;
	    gotLf = TRUE;
	    break;
	    }
	}
    if (!gotLf && bytesInBuf == lf->bufSize)
        {
	if (bufSize > 1024*1024)
	    {
	    errAbort("Line too long (more than %d chars) line %d of %s",
		lf->bufSize, lf->lineIx+1, lf->fileName);
	    }
	else
	    {
	    lineFileExpandBuf(lf, bufSize*2);
	    buf = lf->buf;
	    }
	}
    }

if (lf->zTerm)
    {
    buf[endIx-1] = 0;
    }
lf->lineStart = newStart = lf->lineEnd;
lf->lineEnd = endIx;
++lf->lineIx;
if (retSize != NULL)
    *retSize = endIx - newStart;
*retStart = buf + newStart;
return TRUE;
}

void lineFileNeedNext(struct lineFile *lf, char **retStart, int *retSize)
/* Fetch next line from file.  Squawk and die if it's not there. */
{
if (!lineFileNext(lf, retStart, retSize))
    errAbort("Unexpected end of file in %s", lf->fileName);
}

void lineFileClose(struct lineFile **pLf)
/* Close up a line file. */
{
struct lineFile *lf;
if ((lf = *pLf) != NULL)
    {
    if (lf->fd > 0 && lf->fd != fileno(stdin))
	close(lf->fd);
    freeMem(lf->fileName);
    freeMem(lf->buf);
    freez(pLf);
    }
}

void lineFileExpectWords(struct lineFile *lf, int expecting, int got)
/* Check line has right number of words. */
{
if (expecting != got)
    errAbort("Expecting %d words line %d of %s got %d", 
	    expecting, lf->lineIx, lf->fileName, got);
}

int lineFileChopNext(struct lineFile *lf, char *words[], int maxWords)
/* Return next non-blank line that doesn't start with '#' chopped into words. */
{
int lineSize, wordCount;
char *line;

while (lineFileNext(lf, &line, &lineSize))
    {
    if (line[0] == '#')
        continue;
    wordCount = chopByWhite(line, words, maxWords);
    if (wordCount != 0)
        return wordCount;
    }
return 0;
}

boolean lineFileNextRow(struct lineFile *lf, char *words[], int wordCount)
/* Return next non-blank line that doesn't start with '#' chopped into words.
 * Returns FALSE at EOF.  Aborts on error. */
{
int wordsRead;
wordsRead = lineFileChopNext(lf, words, wordCount);
if (wordsRead == 0)
    return FALSE;
if (wordsRead < wordCount)
    lineFileExpectWords(lf, wordCount, wordsRead);
return TRUE;
}

int lineFileNeedNum(struct lineFile *lf, char *words[], int wordIx)
/* Make sure that words[wordIx] is an ascii integer, and return
 * binary representation of it. */
{
char *ascii = words[wordIx];
char c = ascii[0];
if (c != '-' && !isdigit(c))
    errAbort("Expecting number field %d line %d of %s, got %s", 
    	wordIx+1, lf->lineIx, lf->fileName, ascii);
return atoi(ascii);
}

void lineFileSkip(struct lineFile *lf, int lineCount)
/* Skip a number of lines. */
{
int i, lineSize;
char *line;

for (i=0; i<lineCount; ++i)
    {
    if (!lineFileNext(lf, &line, &lineSize))
        errAbort("Premature end of file in %s", lf->fileName);
    }
}
