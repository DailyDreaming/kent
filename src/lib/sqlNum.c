/* sqlnum.c - Routines to convert from ascii to integer
 * representation of numbers. 
 *
 * This file is copyright 2002 Jim Kent, but license is hereby
 * granted for all use - public, private or commercial. */

#include "common.h"
#include "sqlNum.h"

unsigned sqlUnsigned(char *s)
/* Convert series of digits to unsigned integer about
 * twice as fast as atoi (by not having to skip white 
 * space or stop except at the null byte.) */
{
unsigned res = 0;
char *p = s;
char c;

while (((c = *(p++)) >= '0') && (c <= '9'))
    {
    res *= 10;
    res += c - '0';
    }
if (c != '\0')
    errAbort("invalid unsigned number: \"%s\"", s);
return res;
}

int sqlSigned(char *s)
/* Convert string to signed integer.  Unlike atol assumes 
 * all of string is number. */
{
int res = 0;
char *p = s;
char c;

if (*p == '-')
    p++;
while (((c = *(p++)) >= '0') && (c <= '9'))
    {
    res *= 10;
    res += c - '0';
    }
if (c != '\0')
    errAbort("invalid signed number: \"%s\"", s);
if (*s == '-')
    return -res;
else
    return res;
}

off_t sqlOffset(char *s)
/* Convert string to an off_t.  Unlike atol assumes all of string is
 * number. */
{
off_t res = 0;
char *p = s;
char c;

if (*p == '-')
    p++;
while (((c = *(p++)) >= '0') && (c <= '9'))
    {
    res *= 10;
    res += c - '0';
    }
if (c != '\0')
    errAbort("invalid signed number: \"%s\"", s);
if (*s == '-')
    return -res;
else
    return res;
}
