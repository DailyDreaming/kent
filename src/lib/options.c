/* Options.c - stuff to handle command line options.
 * This is smaller and more flexible than the cgiSpoof
 * routines we used to use - though cgiSpoof is still the
 * method of choice for actual CGI routine. */

#include "common.h"
#include "hash.h"
#include "options.h"

struct hash *optionParseIntoHash(int *pArgc, char *argv[], boolean justFirst)
/* Read options in command line (only up to first real argument) into
 * options hash.   Options come in three forms:
 *      -option         words starting with dash
 *      option=val      words with = in the middle
 *      -option=val     combining the two.
 * The resulting hash will be keyed by the option name with the val
 * string for value.  For '-option' types the value is 'on'. */
{
int i, origArgc, curArgc;
char **rdPt = argv+1, **wrPt = argv+1;
boolean gotReal = FALSE;
struct hash *hash = newHash(6);

origArgc = curArgc = *pArgc;

for (i=1; i<origArgc; ++i)
    {
    boolean startDash;
    if (gotReal)
        *wrPt++ = *rdPt++;
    else
        {
	char *name = *rdPt++;
	boolean startDash = (name[0] == '-');
	char *eqPos = strchr(name, '=');
	if (startDash || eqPos != NULL)
	    {
	    char *val;
	    if (startDash)
	        ++name;
	    if (eqPos != NULL)
	        {
		val = eqPos+1;
		*eqPos = 0;
		}
	    else
	        val = "on";
	    hashAdd(hash, name, val);
	    curArgc -= 1;
	    }
	else
	    {
	    *wrPt++ = name;
	    if (justFirst)
		gotReal = TRUE;
	    }
	}
    }
*pArgc = curArgc;
return hash;
}

static struct hash *options;

void optionHashSome(int *pArgc, char *argv[], boolean justFirst)
/* Set up option hash from command line, optionally only adding
 * up to first non-optional word. */
{
if (options == NULL)
    options = optionParseIntoHash(pArgc, argv, justFirst);
}

void optionHash(int *pArgc, char *argv[])
/* Read options in command line into options hash.   
 * Options come in three forms:
 *      -option         words starting with dash
 *      option=val      words with = in the middle
 *      -option=val     combining the two.
 * The resulting hash will be keyed by the option name with the val
 * string for value.  For '-option' types the value is 'on'. */
{
optionHashSome(pArgc, argv, FALSE);
}

static char *optGet(char *name)
/* Lookup option name.  Complain if options hash not set. */
{
if (options == NULL)
    errAbort("optGet called before optionHash");
return hashFindVal(options, name);
}
 
char *optionVal(char *name, char *defaultVal)
/* Return named option if in options hash, otherwise default. */
{
char *ret = optGet(name);
if (ret == NULL)
     ret = defaultVal;
return ret;
}

int optionInt(char *name, int defaultVal)
/* Return integer value of named option, or default value
 * if not set. */
{
char *s = optGet(name);
if (s == NULL)
    return defaultVal;
if ((s[0] == '-' && isdigit(s[1])) || isdigit(s[0]))
    return atoi(s);
errAbort("option %s has to be integer valued", name);
return 0;
}

float optionFloat(char *name, float defaultVal)
/* Return floating point value or default value if not set. */
{
char *s = optGet(name);
if (s == NULL)
    return defaultVal;
if ((s[0] == '-' && isdigit(s[1])) || isdigit(s[0]))
    return atof(s);
errAbort("option %s has to be float valued", name);
return 0;
}

boolean optionExists(char *name)
/* Return TRUE if option has been set. */
{
return optGet(name) != NULL;
}


