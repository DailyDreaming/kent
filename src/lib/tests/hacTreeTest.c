/* hacTreeTest - Read in items from a file and print the resulting clusters. */
#include "common.h"
#include "linefile.h"
#include "options.h"
#include "sqlNum.h"
#include "hacTree.h"

void usage()
/* Explain usage and exit. */
{
errAbort(
  "hacTreeTest - Read in lines from a file and print the resulting clusters by length.\n"
  "usage:\n"
  "   hacTreeTest in.txt out.txt\n"
  "Actually, before clustering in.txt, it clusters some numbers because that is easier.\n"
  "\n"
  );
}

static struct optionSpec options[] = {
   {NULL, 0},
};

//------------------- numeric example ------------------

static void rPrintSlDoubleTree(FILE *f, struct hacTree *tree, int level)
/* Recursively print out cluster as nested-parens with {}'s around leaf nodes. */
{
double val = ((struct slDouble *)(tree->itemOrCluster))->val;
int i;
for (i=0;  i < level;  i++)
    fputc(' ', f);
if (tree->left == NULL && tree->right == NULL)
    {
    fprintf(f, "{%0.3lf}", val);
    return;
    }
else if (tree->left == NULL || tree->right == NULL)
    errAbort("\nHow did we get a node with one NULL kid??");
fprintf(f, "(%0.3lf:%0.3lf:\n", val, tree->childDistance);
rPrintSlDoubleTree(f, tree->left, level+1);
fputs(",\n", f);
rPrintSlDoubleTree(f, tree->right, level+1);
fputc('\n', f);
for (i=0;  i < level;  i++)
    fputc(' ', f);
fputs(")", f);
}

void printSlDoubleTree(FILE *f, struct hacTree *tree)
/* Print out cluster as nested-parens with {}'s around leaf nodes. */
{
if (tree == NULL)
    {
    fputs("Empty tree.\n", f);
    return;
    }
rPrintSlDoubleTree(f, tree, 0);
fputc('\n', f);
}

double slDoubleDistance(const struct slList *item1, const struct slList *item2, void *extraData)
/* Return the absolute difference between the two kids' values. */
{
const struct slDouble *kid1 = (const struct slDouble *)item1;
const struct slDouble *kid2 = (const struct slDouble *)item2;
double diff = kid1->val - kid2->val;
if (diff < 0)
    diff = -diff;
return diff;
}

struct slList *slDoubleMidpoint(const struct slList *item1, const struct slList *item2,
				void *unusedExtraData)
/* Make a new slDouble that is the midpoint/average of kids' values. */
{
const struct slDouble *kid1 = (const struct slDouble *)item1;
const struct slDouble *kid2 = (const struct slDouble *)item2;
double midpoint = (kid1->val + kid2->val) / 2.0;
return (struct slList *)(slDoubleNew(midpoint));
}

void hacTreeTestInts(FILE *f, struct lm *localMem)
/* Cluster a list of integers (use doubles to get more precise cluster points). */
{
struct slDouble *sldList = NULL;
slAddHead(&sldList, slDoubleNew(0));
slAddHead(&sldList, slDoubleNew(5));
slAddHead(&sldList, slDoubleNew(7));
slAddHead(&sldList, slDoubleNew(1));
slAddHead(&sldList, slDoubleNew(-8));
slAddHead(&sldList, slDoubleNew(12));
slAddHead(&sldList, slDoubleNew(6));
slAddHead(&sldList, slDoubleNew(-6));
slAddHead(&sldList, slDoubleNew(-3));
slAddHead(&sldList, slDoubleNew(8));
struct hacTree *clusters = hacTreeFromItems((struct slList *)sldList, localMem,
					    slDoubleDistance, slDoubleMidpoint, NULL);
fputs("Clustering by numeric value:\n", f);
printSlDoubleTree(f, clusters);
}


//------------------- strlen(line) example ------------------

static void rPrintSlNameTree(FILE *f, struct hacTree *tree, int level)
/* Recursively print out cluster as nested-parens with {}'s around leaf nodes. */
{
char *contents = ((struct slName *)(tree->itemOrCluster))->name;
int i;
for (i=0;  i < level;  i++)
    fputc(' ', f);
if (tree->left == NULL && tree->right == NULL)
    {
    fprintf(f, "{%s}", contents);
    return;
    }
else if (tree->left == NULL || tree->right == NULL)
    errAbort("\nHow did we get a node with one NULL kid??");
fprintf(f, "(%s:%d:\n", contents, (int)(tree->childDistance));
rPrintSlNameTree(f, tree->left, level+1);
fputs(",\n", f);
rPrintSlNameTree(f, tree->right, level+1);
fputc('\n', f);
for (i=0;  i < level;  i++)
    fputc(' ', f);
fputs(")", f);
}

void printSlNameTree(FILE *f, struct hacTree *tree)
/* Print out cluster as nested-parens with {}'s around leaf nodes. */
{
if (tree == NULL)
    {
    fputs("Empty tree.\n", f);
    return;
    }
rPrintSlNameTree(f, tree, 0);
fputc('\n', f);
}

double strlenDistance(const struct slList *item1, const struct slList *item2, void *extraData)
/* Return the absolute difference in length between the two kids. */
{
const struct slName *kid1 = (const struct slName *)item1;
const struct slName *kid2 = (const struct slName *)item2;
int diff = strlen(kid1->name) - strlen(kid2->name);
if (diff < 0)
    diff = -diff;
return (double)diff;
}

struct slList *mergeLines(const struct slList *item1, const struct slList *item2,
			  void *unusedExtraData)
/* Make a string that is the average length of the kids. */
{
const struct slName *kid1 = (const struct slName *)item1;
const struct slName *kid2 = (const struct slName *)item2;
int len1 = strlen(kid1->name);
int len2 = strlen(kid2->name);
int avgLen = (len1 + len2 + 1) / 2;
const char *longerName = (len1 >= len2) ? kid1->name : kid2->name;
return (struct slList *)slNameNewN((char *)longerName, avgLen);
}

void hacTreeTestStrlens(char *inFileName, FILE *f, struct lm *localMem)
/* Read in items from a file and print the resulting clusters. */
{
struct slName *sln, *slnList = NULL;
struct lineFile *lf = lineFileOpen(inFileName, TRUE);
char *line;
int size;
while (lineFileNext(lf, &line, &size))
    {
    sln = slNameNewN(line, size);
    slAddHead(&slnList, sln);
    }
lineFileClose(&lf);
slReverse(&slnList);
struct hacTree *clusters = hacTreeFromItems((struct slList *)slnList, localMem,
					    strlenDistance, mergeLines, NULL);
fputs("Clustering by string length:\n", f);
printSlNameTree(f, clusters);
carefulClose(&f);
}

void hacTreeTest(char *inFileName, char *outFileName)
/* Read in items from a file and print the resulting clusters. */
{
FILE *f = mustOpen(outFileName, "w");
struct lm *localMem = lmInit(0);
hacTreeTestInts(f, localMem);
hacTreeTestStrlens(inFileName, f, localMem);
lmCleanup(&localMem);
}

int main(int argc, char *argv[])
/* Process command line. */
{
optionInit(&argc, argv, options);
if (argc != 3)
    usage();
hacTreeTest(argv[1], argv[2]);
return 0;
}
