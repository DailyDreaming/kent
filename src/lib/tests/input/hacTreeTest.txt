one
there it is
two
is there
is it
there it was
three
