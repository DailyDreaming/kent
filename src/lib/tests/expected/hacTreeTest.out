Clustering by numeric value:
(2.562:13.375:
 (-4.125:5.750:
  (-1.250:3.500:
   (0.500:1.000:
    {1.000},
    {0.000}
   ),
   {-3.000}
  ),
  (-7.000:2.000:
   {-6.000},
   {-8.000}
  )
 ),
 (9.250:5.500:
  (6.500:2.000:
   (5.500:1.000:
    {6.000},
    {5.000}
   ),
   (7.500:1.000:
    {8.000},
    {7.000}
   )
  ),
  {12.000}
 )
)
Clustering by string length:
(there i:6:
 (there it w:4:
  (there it was:1:
   {there it is},
   {there it was}
  ),
  {is there}
 ),
 (is i:2:
  (is it:0:
   {is it},
   {three}
  ),
  (one:0:
   {one},
   {two}
  )
 )
)
